### ISSUE TITLE
AllValidator should not check for type

### ISSUE THREAD

**Issue by zerkms** [symfony/symfony#26463]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | yes
| RFC?             | no
| Symfony version  | 4.0.6

At the moment the `AllValidator` makes the following extra check: 

```php
        if (!is_array($value) && !$value instanceof \Traversable) {
            throw new UnexpectedTypeException($value, 'array or Traversable');
        }
```
it causes a problem to validate whether a value is an array containing only integers.

Let's see the problem in action. Let's say we have an object that has a field `$ids` that must be a non-empty array of integers. The naive attempt would be to annotate it like this:

```php
    /**
     * @var int[]
     * @Assert\NotBlank()
     * @Assert\Type("array")
     * @Assert\All({
     *     @Assert\Type("int"),
     *     @Assert\NotBlank(),
     * })
     */
    private $ids;
```
It works fine, as soon as you pass an array there. But if you pass non-array and non-iterable value, like `42` it fails miserably with the following exception:

```
Expected argument of type "array or Traversable", "integer" given
```

Why does the `AllValidator` throws at all?

```php
        if (null === $value) {
            return;
        }
```

here it checks whether a value is set at all - so that you could validate it with `NotBlank` if necessary.

So why does it do the type check with exception, while that part of the job should have been done a `Type("array")` constraint instead?

What I suggest:

```php
        if (!is_array($value) && !$value instanceof \Traversable) {
            return;
        }
```
instead. In that case if the value is not an array or traversable - the validation does not run. And if one wants to ensure it's an array - they must do a naive `Type("array")` check before.

**Comment by xabbuh** [comment]
Are you sure that the `Type` constraint does not add a violation?

**Comment by zerkms** [comment]
@xabbuh it matters in what order they invoked: if `Type` is checked first - it would be added (I haven't checked it but I see no reason why it wouldn't).

**Comment by xabbuh** [comment]
In that case I don't think that this is a bug. If you cannot ensure that the value you check using composite constraints is a collection, I suggest to add the `Type` constraint as you did, but additionally make use of validation groups and a group sequence which checks types first and only executes the other checks when the type requirements are fulfilled.

**Comment by zerkms** [comment]
@xabbuh added `Type` validation changes nothing - since the exception is still thrown. At the moment `All` validator does more job than it should have - it simply MUST NOT check for the type, it's not its responsibility.

At the moment it's impossible to validate collections with symfony validator.

I honestly cannot see what benefits that type check brings.

**Comment by zerkms** [comment]
To clarify why it's a bug:

1. If you pass an array - then validation works as expected
2. If you pass not an array (eg a number) - then the validator throws an exception (while it should return a validation violation)

The `All` type does not check for emptiness - it's a `NotBlank` validator responsibility. The same way it must not check for the type - it's a `Type` validator responsibility.

**Comment by zerkms** [comment]
Right, I checked other constraints and it looks like almost every other constraint check for type. Which makes symfony/validator impossible to validate an arbitrary input. That's correct: validator must be able to validate anything, but it cannot. Correct me somebody if I wrong.
